### PR TITLE
Don't misdetect circular deps because of incorrect dependencies

### PR DESCRIPTION
**Pull request by ted-gould**
Started with the problem in turtlesim that randomly it would detect a circular dependency issue, and break the build. Wanted to track that down so I wrote a couple of tests to recreate. Turns out it was because we were sometimes populating the dependencies incorrectly. Fixed that.

### CODE REVIEW

File: snapcraft/yaml.py [hunk 1]
```diff
@@ -129,6 +129,8 @@ def __init__(self):
 
             if 'after' in properties:
                 after_requests[part_name] = properties.pop('after')
+                if isinstance(after_requests[part_name], str):
```
Comment (sergiusens) [inline snapcraft/yaml.py]: nice catch!

How about a unit test for this part of the logic?

File: integration-tests/data/simple-circle/tree/snapcraft.yaml [hunk 2]
```diff
@@ -0,0 +1,92 @@
+name: dep-tree
+summary: A tree of things to make sure work
+description: These don't really make sense for tests
```
Comment (sergiusens) [inline integration-tests/data/simple-circle/tree/snapcraft.yaml]: Add a longer description here for the test (comment I once got from @elopio) like

```yaml
description: |
  The objective of this test is to blah ...
  more description ... blah blah
```

File: integration-tests/data/simple-circle/circle/snapcraft.yaml [hunk 3]
```diff
@@ -0,0 +1,53 @@
+name: real-circle
+summary: A big bad circle
+description: These don't really make sense for tests
```
Comment (sergiusens) [inline integration-tests/data/simple-circle/circle/snapcraft.yaml]: Add a longer description here for the test (comment I once got from @elopio) like

```yaml
description: |
  The objective of this test is to blah ...
  more description ... blah blah

**Comment by ted-gould** [comment]
Beat that! All green. /me drops mic

**Comment by elopio** [comment]
Wow, even coveralls is happy with you. I'm impressed :+1:

File: snapcraft/yaml.py [hunk 4]
```diff
@@ -129,6 +129,8 @@ def __init__(self):
...
+                    after_requests[part_name] = [after_requests[part_name]]
```
Comment (elopio) [inline snapcraft/yaml.py]: This is a little unreadable. You are replacing the var with a list of itself.
I would either add a comment or try to a make it clearer what's going on. What do you think of this?
after_requests[part_name] = []
after_requests.extend(properties.pop('after'))

File: snapcraft/yaml.py [hunk 5]
```diff
@@ -128,7 +128,11 @@ def __init__(self):
                         plugin_name))
 
             if 'after' in properties:
-                after_requests[part_name] = properties.pop('after')
+                afterprop = properties.pop('after')
+                if isinstance(afterprop, str):
```
Comment (sergiusens) [inline snapcraft/yaml.py]: btw, when is after ever non array? The schema should capture this :-/

Comment (ted-gould) [inline snapcraft/yaml.py]: There is no schema for these properties. I checked there first. Added a deprecation warning as I think it's very typically used as a string today.

File: snapcraft/yaml.py [hunk 6]
```diff
@@ -128,7 +128,14 @@ def __init__(self):
...
+                    logger.warning(
+                        'DEPRECATED: after on {} should be a list of strings '
+                        'not a string itself.'.format(plugin_name))
+                    after_requests[part_name] = [afterprop]
+                else:
+                    after_requests[part_name] = afterprop
```
Comment (elopio) [inline snapcraft/yaml.py]: +1 from me.

File: snapcraft/yaml.py (hunk 5 above)
Comment (sergiusens) [inline snapcraft/yaml.py]: Just to record this here, I think we should just implement this correctly, given that we iterated over the string contents before I don't see how this worked for anyone so it is a plain bug and requires no deprecation message.

**Comment by sergiusens** [comment]
I'm closing the PR for book keeping and to avoid an accidental merge.
Thanks for working on this.

**Comment by ted-gould** [comment]
So I had just continued with this branch... can make a new one. But, eh, I don't see a reason.

File: snapcraft/lifecycle.py [hunk 7]
```diff
@@ -371,3 +356,73 @@ def _validate_relative_paths(files):
...
             raise PluginError('path "{}" must be relative'.format(d))
+
+
+def _add_required_schema(plugin_schema):
+    if plugin_schema:
+        schema = plugin_schema
+        if 'properties' not in schema:
+            schema['properties'] = {}
+    else:
+        schema = {
```
Comment (sergiusens) [inline snapcraft/lifecycle.py]: why don't you load `schema/snapcraft.yaml` which is already loaded in `yaml.py`, so you can just pass the relevant schema part for properties defined in https://github.com/ubuntu-core/snapcraft/blob/master/schema/snapcraft.yaml#L97 and also add the missing `after`, `organize`, ... , there.

**Comment by sergiusens** [comment]
Thank you, especially for the patience in all the back on forth that this involved.

**Pull request merged**